src/Microsoft.AspNetCore.Http.Abstractions/Routing/Endpoint.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Routing
+{
+    /// <summary>
+    /// Respresents a logical endpoint in an application.
+    /// </summary>
+    public abstract class Endpoint
```

Comment:
What about:

```
public interface IEndpoint 
{ 
    EndpointMetadataCollection Metadata { get; } 
}
```

cc @rynowak

Reply:
This spawned an IM discussion. I've tried a few different approaches and the way things are is the best compromise I've found.


Goals:

I don't want Endpoint/IEndpoint to become something that application/framework authors subclass. I want it to be **data** not a service. The world breaks when these becomes services. I also don't want it to be polymorphic because we are creating a builder API to create these. Builder APIs get wierd when when the thing you're buildering is polymorphic. (I can give examples if necessary).

I don't want to leak all of https://github.com/aspnet/Routing/tree/release/2.2/src/Microsoft.AspNetCore.Routing/Patterns into this repo. I think it's important that the primary type (`MatcherEndpoint`) exposes details like the route pattern as a first-class property, accepts it in the constructor, etc.

The separation largely enforces the layering we already know we want. If you want to see and understand what routing did in a middleware, you just need `Endpoint`. If you want to register things in routing, you need to reference routing to use `MatcherEndpoint`. It doesn't make much sense to move `MatcherEndpoint` down to this level, because you wouldn't be able to use them for anything without routing.

Reply:
+1 that it is data and not a behavior to bolt onto other classes.

Reply:
I had a long discussion with @rynowak and I understand how we landed here but there are things that give me pause. This doesn't feel like an Endpoint, it just feels like something that gives you metadata. The Endpoint is really the `MatcherEndpoint`, that's the thing everyone will create. This type is going to be used by middleware and will hang off the IEndpointFeature. 

It feels more like a container for metadata than anything else.

Reply:
> It feels more like a container for metadata than anything else.

If we moved the request delegate down from `MatcherEndpoint` to here, would that address your concerns?